src/Microsoft.Identity.Web.DownstreamApi/DownstreamApi.cs

```diff
@@ -286,15 +287,45 @@ internal partial class DownstreamApi : IDownstreamApi
             // Downstream API URI
             string apiUrl = effectiveOptions.GetApiUrl();
 
-            // Creation of the Http Request message with the right HTTP Method
-            using HttpRequestMessage httpRequestMessage = new HttpRequestMessage(
+            using HttpClient client = string.IsNullOrEmpty(serviceName) ? _httpClientFactory.CreateClient() : _httpClientFactory.CreateClient(serviceName);
+
+            // Create an HTTP request message
+            using HttpRequestMessage httpRequestMessage = new(
                 new HttpMethod(effectiveOptions.HttpMethod),
                 apiUrl);
-            if (content != null)
-            {
-                httpRequestMessage.Content = content;
-            }
 
+            await UpdateRequestAsync(httpRequestMessage, content, effectiveOptions, appToken, user, cancellationToken);
+
+            // Send the HTTP message           
+            var downstreamApiResult = await client.SendAsync(httpRequestMessage, cancellationToken).ConfigureAwait(false);
+
+            effectiveOptions.AcquireTokenOptions.Claims = WwwAuthenticateParameters.GetClaimChallengeFromResponseHeaders(downstreamApiResult.Headers);
+
+            // Retry if the resource sent 401 Unauthorized with WWW-Authenticate header and claims
+            if (downstreamApiResult.StatusCode == System.Net.HttpStatusCode.Unauthorized && !string.IsNullOrEmpty(effectiveOptions.AcquireTokenOptions.Claims))
+            {
+                using HttpRequestMessage retryHttpRequestMessage = new(
+                    new HttpMethod(effectiveOptions.HttpMethod),
+                    apiUrl);
+
+                await UpdateRequestAsync(retryHttpRequestMessage, content, effectiveOptions, appToken, user, cancellationToken);
+
+                return await client.SendAsync(retryHttpRequestMessage, cancellationToken).ConfigureAwait(false);
+            }
+
+            return downstreamApiResult;
+        }
+
+        private async Task UpdateRequestAsync(
+            HttpRequestMessage request,
+            HttpContent? content,
+            DownstreamApiOptions effectiveOptions,
+            bool appToken,
+            ClaimsPrincipal? user,
+            CancellationToken cancellationToken)
+        {
+            request.Content = content;

```

Comment:
can we do that if the content is null? (which will happen for HTTP Get for instance?